src/irr/asset/CGLSLBRDFBuiltinIncludeLoader.h

```diff
@@ -1,18 +1,209 @@
+
+// do not use this struct in SSBO or UBO, its wasteful on memory
+struct irr_glsl_BSDFIsotropicParams
+{
+   float NdotL;
+   float NdotL_squared;
+   float NdotV;
+   float NdotV_squared;
+   float VdotL; // same as LdotV
+   float NdotH;
+   float VdotH; // same as LdotH
+   // left over for anisotropic calc and BSDF that want to implement fast bump mapping
+   float LplusV_rcpLen;
+   // basically metadata
+   vec3 L;
+   irr_glsl_ViewSurfaceInteraction interaction;
+};
+
+// do not use this struct in SSBO or UBO, its wasteful on memory
+struct irr_glsl_BSDFAnisotropicParams
+{
+   irr_glsl_BSDFIsotropicParams isotropic;
+   float TdotL;
+   float TdotV;
+   float TdotH;
+   float BdotL;
+   float BdotV;
+   float BdotH;
+   // useless metadata
+   vec3 T;
+   vec3 B;
+};
+
+/*
+//TODO error no operator '*' for vec3 and mat2x3
+// chain rule on various functions (usually vertex attributes and barycentrics)
+vec2 irr_glsl_applyScreenSpaceChainRule1D3(in vec3 dFdG, in mat2x3 dGdScreen)
+{
+   return dFdG*dGdScreen;
+}
+mat2 irr_glsl_applyScreenSpaceChainRule2D3(in mat2x3 dFdG, in mat2 dGdScreen)
+{
+   return dFdG*dGdScreen;
+}
+mat2x3 irr_glsl_applyScreenSpaceChainRule3D3(in mat3 dFdG, in mat2x3 dGdScreen)
+{
+   return dFdG*dGdScreen;
+}
+mat2x4 irr_glsl_applyScreenSpaceChainRule4D3(in mat3x4 dFdG, in mat2x3 dGdScreen)
+{
+   return dFdG*dGdScreen;
+}
+*/
+
+// only in the fragment shader we have access to implicit derivatives
+irr_glsl_ViewSurfaceInteraction irr_glsl_calcFragmentShaderSurfaceInteraction(in vec3 _CamPos, in vec3 _SurfacePos, in vec3 _Normal)
+{
+   irr_glsl_ViewSurfaceInteraction interaction;
+   interaction.V.dir = _CamPos-_SurfacePos;
+   interaction.V.dPosdScreen[0] = dFdx(_SurfacePos);
+   interaction.V.dPosdScreen[1] = dFdy(_SurfacePos);
+   interaction.N = _Normal;
+   return interaction;
+}
+/*
+//TODO it doesnt compile, lots of undefined symbols
+// when you know the projected positions of your triangles (TODO: should probably make a function like this that also computes barycentrics)
+irr_glsl_ViewSurfaceInteraction irr_glsl_calcBarycentricSurfaceInteraction(in vec3 _CamPos, in vec3 _SurfacePos[3], in vec3 _Normal[3], in float _Barycentrics[2], in vec2 _ProjectedPos[3])
+{
+   irr_glsl_ViewSurfaceInteraction interaction;
+
+   // Barycentric interpolation = b0*attr0+b1*attr1+attr2*(1-b0-b1)
+   vec3 b = vec3(_Barycentrics[0],_Barycentrics[1],1.0-_Barycentrics[0]-_Barycentrics[1]);
+   mat3 vertexAttrMatrix = mat3(_SurfacePos[0],_SurfacePos[1],_SurfacePos[2]);
+   interaction.V.dir = _CamPos-vertexAttrMatrix*b;
+   // Schied's derivation - modified
+   vec2 to2 = _ProjectedPos[2]-_ProjectedPos[1];
+   vec2 to1 = _ProjectedPos[0]-_ProjectedPos[1];
+   float d = 1.0/determinant(mat2(to2,to1)); // TODO double check all this
+   mat2x3 dBaryd = mat2x3(vec3(v[1].y-v[2].y,to2.y,to0.y)*d,-vec3(v[1].x-v[2].x,to2.x,t0.x)*d);
+   //
+   interaction.dPosdScreen = irr_glsl_applyScreenSpaceChainRule3D3(vertexAttrMatrix,dBaryd);
```

Comment:
thats why it doesnt compile cause that function is not defined (commented out) XD

Reply:
no, i meant this function doesnt  compile. The shader compiles BECAUSE it's commented out. It's not used anyway but shaderc still tries to compile it

Reply:
well you need to uncomment irr_glsl_applyScreenSpaceChainRule3D3 and define it and it will work.